daffodil-core/src/main/scala/edu/illinois/ncsa/daffodil/grammar/ElementBaseGrammarMixin.scala

```diff
@@ -738,16 +770,25 @@ trait ElementBaseGrammarMixin
       }
 
       case PrimType.Decimal => {
-        Assert.invariant(binaryIntRep == bin)
         if (binaryDecimalVirtualPoint > tunable.maxBinaryDecimalVirtualPoint)
           SDE("Property binaryDecimalVirtualPoint %s is greater than limit %s", binaryDecimalVirtualPoint, tunable.maxBinaryDecimalVirtualPoint)
         if (binaryDecimalVirtualPoint < tunable.minBinaryDecimalVirtualPoint)
           SDE("Property binaryDecimalVirtualPoint %s is less than limit %s", binaryDecimalVirtualPoint, tunable.minBinaryDecimalVirtualPoint)
         if (binaryNumberKnownLengthInBits == -1 ||
           binaryNumberKnownLengthInBits > 8) byteOrderRaw // must have or SDE
-        binaryNumberKnownLengthInBits match {
-          case -1 => new BinaryDecimalRuntimeLength(this)
-          case _ => new BinaryDecimalKnownLength(this, binaryNumberKnownLengthInBits)
+
+        (binaryNumberRep, lengthKind, binaryNumberKnownLengthInBits) match {
+          case (BinaryNumberRep.Binary, _, -1) => new BinaryDecimalRuntimeLength(this)
+          case (BinaryNumberRep.Binary, _, _) => new BinaryDecimalKnownLength(this, binaryNumberKnownLengthInBits)
+          case (BinaryNumberRep.Packed, LengthKind.Delimited, -1) => new PackedDecimalDelimitedEndOfData(this)
+          case (BinaryNumberRep.Packed, _, -1) => new PackedDecimalRuntimeLength(this)
+          case (BinaryNumberRep.Packed, _, _) => new PackedDecimalKnownLength(this, binaryNumberKnownLengthInBits)
+          case (BinaryNumberRep.Bcd, LengthKind.Delimited, -1) => new BCDDecimalDelimitedEndOfData(this)
+          case (BinaryNumberRep.Bcd, _, -1) => new BCDDecimalRuntimeLength(this)
+          case (BinaryNumberRep.Bcd, _, _) => new BCDDecimalKnownLength(this, binaryNumberKnownLengthInBits)
+          case (BinaryNumberRep.Ibm4690Packed, LengthKind.Delimited, -1) => new IBM4690PackedDecimalDelimitedEndOfData(this)
+          case (BinaryNumberRep.Ibm4690Packed, _, -1) => new IBM4690PackedDecimalRuntimeLength(this)
+          case (BinaryNumberRep.Ibm4690Packed, _, _) => new IBM4690PackedDecimalKnownLength(this, binaryNumberKnownLengthInBits)
```

Comment:
Related to the grammar, Section 12.1.1 of the spec says that implicit alignment of packed decimals is always 8. See getImplicitAlignmentInBits in ElementBase.scala. That needs to be modified to check the binary rep.

Reply:
Also related to the grammar, Section 12.3.3 says that implicit lengths are not allowed for packed decimals. Probably binaryNumberKnownLengthInBits needs an SDE based on the binary rep if lengthKind is implicit.

Reply:
Also related to the grammar, Section 12.1.3 says packed decimal data must have alignment that is a multiple of 4 and is a SDE otherwise.

Reply:
Section 12.3.7.2.3 says it is a schema definition error if the specified length is not a multiple of 4 bits. Need to check that at schema compilation time for constant value and at runtime for runtime values.

Reply:
Checks and tests for these edge cases have been added in my most recent commit 6d3a01f8c0b68428c063726a62cf2d25a397fd55